flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesJobGraphStoreWatcher.java

```diff
@@ -0,0 +1,148 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.kubernetes.utils.Constants.JOB_GRAPH_STORE_KEY_PREFIX;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link JobGraphStoreWatcher} implementation for Kubernetes. It watch the Dispatcher leader ConfigMap and call the
+ * {@link JobGraphStore.JobGraphListener} based on the received event.
+ */
+public class KubernetesJobGraphStoreWatcher implements JobGraphStoreWatcher {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesJobGraphStoreWatcher.class);
+
+	private final FlinkKubeClient kubeClient;
+
+	private final String configMapName;
+
+	private JobGraphStore.JobGraphListener jobGraphListener;
+
+	@Nullable
+	private KubernetesWatch kubernetesWatch;
+
+	public KubernetesJobGraphStoreWatcher(FlinkKubeClient kubeClient, String configMapName) {
+		this.kubeClient = checkNotNull(kubeClient);
+		this.configMapName = checkNotNull(configMapName);
+	}
+
+	@Override
+	public void start(JobGraphStore.JobGraphListener jobGraphListener) {
+		this.jobGraphListener = checkNotNull(jobGraphListener);
+		kubernetesWatch = kubeClient.watchConfigMaps(configMapName, new ConfigMapCallbackHandlerImpl());
+	}
+
+	@Override
+	public void stop() {
+		if (kubernetesWatch != null) {
+			kubernetesWatch.close();
+		}
+	}
+
+	private class ConfigMapCallbackHandlerImpl implements FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> {
+		// This is used to get the difference between current and previous data. And then notify the listener.
+		final Set<JobID> previousJobIDs = new HashSet<>();
+
+		@Override
+		public void onAdded(List<KubernetesConfigMap> configMaps) {
+			// The ConfigMap is created by KubernetesLeaderElectionDriver with empty data. We do not process this
+			// useless event.
+		}
+
+		@Override
+		public void onModified(List<KubernetesConfigMap> configMaps) {
+			handleConfigMapEvent(configMaps);
+		}
+
+		@Override
+		public void onDeleted(List<KubernetesConfigMap> configMaps) {
+			// The ConfigMap will be created again in the leader election service.
+		}
+
+		@Override
+		public void onError(List<KubernetesConfigMap> configMaps) {
+			LOG.error("Error while watching the configMap {}", configMapName);
+		}
+
+		@Override
+		public void handleFatalError(Throwable throwable) {
+			LOG.error("Error while watching the configMap {}", configMapName, throwable);
+		}
+
+		private void handleConfigMapEvent(List<KubernetesConfigMap> configMaps) {
+			final KubernetesConfigMap configMap = KubernetesUtils.checkConfigMaps(configMaps, configMapName);
+			final Set<JobID> currentJobIDs = getJobIDs(configMap);
+			notifyJobGraphListenChanges(currentJobIDs, previousJobIDs);
+			previousJobIDs.clear();
+			previousJobIDs.addAll(currentJobIDs);
```

Comment:
After double-check with other Kubernetes experts, I got the confirmation that for each watcher on specific resource(e.g. ConfigMap) here, Kubernetes API server could guarantee that the events are sent in order. Since we are using a web-socket for the watcher, which is a long connection, I think we could receive the events in correct order.

Reply:
Could we make an assertion here? E.g., checking the resource version. Because we are highly depending on the order in which events are received.

Reply:
I will add a `checkState` here.